torch/jit/__init__.py

```diff
@@ -900,10 +900,65 @@ def _try_compile_weak_script(fn):
         _jit_internal.compiled_weak_fns[fn]["compiled_fn"] = compiled_fn
         entry["status"] = _jit_internal.COMPILED
         return compiled_fn
+        # TODO: use fn.__closure__
+        raise RuntimeError("Cannot make resolutionCallback in Python 2")
     else:
         return entry["compiled_fn"]
 
 
+class ScriptWarning(Warning):
+    pass
+
+
+def make_rcb(fn):
+    """
+    Create a resolutionCallback by introspecting the function instead of
+    looking up the stack for the enclosing scope
+    """
+    if PY2:
+        # TODO: use fn.__closure__
+        raise RuntimeError("Cannot make resolutionCallback in Python 2")
+    else:
+        closure = inspect.getclosurevars(fn)
+    f_globals = closure.globals
+    f_unbound = closure.unbound
+    f_builtins = closure.builtins
+
+    def env(key):
+        if key in f_globals:
+            return f_globals[key]
+        elif key in f_builtins:
+            return f_builtins[key]
+        elif key in unbound:
+            raise RuntimeError("Cannot resolve unbound variable", key)
+        elif hasattr(builtins, key):
+            return getattr(builtins, key)
+        else:
+            return None
+
+    return env
+
+
+def _try_compile_fn(fn):
+    if inspect.ismethod(fn) or _is_ignored_function(fn):
+        # Skip methods
+        return None
+    try:
+        rcb = make_rcb(fn)
+        return torch.jit.script(fn, _rcb=rcb)
+    except Exception as e:
+        if hasattr(fn, '__qualname__'):
+            name = fn.__qualname__
+        elif hasattr(fn, '__name__'):
+            name = fn.__name__
+        else:
+            name = '<unknown>'
+        # Compile error, display as a warning so the user knows why it didn't
+        # compile and that it was inserted as a Python op
+        message = "Inserting a call to Python for '{}' since it could not be compiled:\n{}".format(name, e)
```

Comment:
We should not have warnings. This should become an error unless it has been annotated now with something like `@torch.jit.as_python`